### ISSUE TITLE
Multiple images

### ISSUE THREAD

**Issue by fakerybakery** [m87-labs/moondream#56]
Moondream2 is incredible! Is it possible to support multiple images? Thanks!

**Comment by ProGamerGov** [comment]
You are referring to a batch size greater than 1 to apply each prompt to, correct?

**Comment by Vigilence** [comment]
Taggui supports moonbeam 1 and 2 as well as batch processing.

**Comment by fakerybakery** [comment]
No, I mean asking one question about multiple images (ie if I have 3 images of animals, and I ask "what do these animals all have in common")

**Comment by Vigilence** [comment]
I see, using several images in a single question/prompt.  That's actually a great idea!

**Comment by fakerybakery** [comment]
I guess technically one could preprocess multiple images into one (ie a grid of images) but the model isn't trained on that and the resolution for each image would be quite low

**Comment by vikhyat** [comment]
Would help to get examples of more real world use-cases for this. I'm definitely open to adding support for it, just need to understand what types of training data to generate.

**Comment by fakerybakery** [comment]
I was mainly thinking about images from multiple perspectives, image choices, etc. For example, what if the security camera example on your homepage could support multiple images? Also would be great for healthcare use-cases

**Comment by zhaohm14** [comment]
Moondream is truly impressive, and your suggestion is fantastic!
There have been times when I wanted to have multi-turn dialogues that need to uploading several images, all while having moondream remember the context and images from prior conversations. For example, it would be incredibly convenient to fluidly discuss different movie elements, like scenes and characters, in an ongoing dialogue.
Hoping to see this feature become a reality!

**Comment by ghost** [comment]
> Would help to get examples of more real world use-cases for this. I'm definitely open to adding support for it, just need to understand what types of training data to generate.

It could be useful for problems involving multiple figure reasoning. For example, finding the differences between two images, or describing small details of a scene where you're given multiple angles, which may help reasoning.

**Comment by shortcipher3** [comment]
I agree, I would like to pas in multiple images and ask if it is the same person in both images or pass in multiple images of a  scene and ask what has changed.  Or pass in multiple images from a property and ask what is happening.  Or pass multiple images from an ADAS system and ask what is happening.

**Comment by saket424** [comment]
I was hoping something like amblegpt (https://github.com/mhaowork/amblegpt) can be supported with a self-hosted moondream backend (as a replacement for a gpt4o backend). 

Thoughts?

```
# Video frame sampling settings
GAP_SECS = 3

# GPT config
DEFAULT_PROMPT = """
You're a helpful assistant helping to label a video for machine learning training
You are reviewing some continuous frames of a video footage as of {EVENT_START_TIME}. Frames are {GAP_SECS} second(s) apart from each other in the chronological order.
{CAMERA_PROMPT}
Please describe what happend in the video in json format. Do not print any markdown syntax!
Answer like the following:
{{
    "num_persons" : 2,
    "persons" : [
    {{
        "height_in_meters": 1.75,
        "duration_of_stay_in_seconds": 15,
        "gender": "female",
        "age": 50
    }},
    {{
        "height_in_meters": 1.60,
        "duration_of_stay_in_seconds": 15,
        "gender": "unknown",
        "age": 36
    }},
    "summary": "SUMMARY"
    "title": "TITLE"
}}

You can guess their height and gender . It is 100 percent fine to be inaccurate.

You can measure their duration of stay given the time gap between frames.

You should take the time of event into account.
For example, if someone is trying to open the door in the middle of the night, it would be suspicious. Be sure to mention it in the SUMMARY.

Mostly importantly, be sure to mention any unusualness considering all the context.

Some example SUMMARIES are
    1. One person walked by towards right corner with her dog without paying attention towards the camera's direction.
    2. One Amazon delivery person (in blue vest) dropped off a package.
    3. A female is waiting, facing the door.
    4. Suspicious: A person is wandering without obvious purpose in the middle of the night, which seems suspicious.
    5. Suspicious: A person walked into the frame from outside, picked up a package, and left.
       The person didn't wear any uniform so this doesn't look like a routine package pickup. Be aware of potential package theft!

TITLE is a one sentence summary of the event. Use no more than 10 words.

Write your answer in {RESULT_LANGUAGE} language.
"""
```

**Comment by beniz** [comment]
> Would help to get examples of more real world use-cases for this. I'm definitely open to adding support for it, just need to understand what types of training data to generate.

hello @vikhyat real-world use case is multimodal RAG in which multiple 'pages' / image chunks from different documents are returned. An approach can be to stitch them together but it's hacky.

**Comment by not-lain** [comment]
cc @fakerybakery 
merging multiple images into a single will result in erroneous generation (for case image is has an x,y axes, the 2 images will merge either along the x axis or the y axis resulting in a single image that the AI will not know where the image starts and where it stops) 

instead i propose we use the same implementation as `HuggingFaceTB/SmolVLM-Instruct` which deep down uses the [Idefics3ForConditionalGeneration](https://github.com/huggingface/transformers/blob/98e8062df3cf82b367e8a243963e4afb0d9d3407/src/transformers/models/idefics3/modeling_idefics3.py#L1047) architecture.
I would also like to invite you to checkout the [inputs_merger](https://github.com/huggingface/transformers/blob/98e8062df3cf82b367e8a243963e4afb0d9d3407/src/transformers/models/idefics3/modeling_idefics3.py#L882) function to understand better how they handled multiple images along with the input tokens.

friendly tagging @vikhyat as well, a moondream version that handles multiple images would surely be a nice addition.

as for the dataset structure, this might be a good one to start with : https://huggingface.co/datasets/HuggingFaceM4/Docmatix

**Comment by KarlisJ** [comment]
> Would help to get examples of more real world use-cases for this. I'm definitely open to adding support for it, just need to understand what types of training data to generate.

More accurate semantic text-to-image search for larger galleries. ATM, the option is to use something like CLIP, save image embeddings and compare cosine distance with text embedding. However, the text has to be very detailed since CLIP is global embedding.

With moondream I can ask if certain image contain X. It would be great to ask entire database which photos contain X.

**Comment by sanggusti** [comment]
Hi, seems like this was already asked a year ago, is there update about this? Seems like from the input type I can only input one image instead of multiple.

**Comment by sanggusti** [comment]
Oh nvm, I just find the closed [issue](https://github.com/vikhyat/moondream/issues/59) about this in particular. Thanks! seems like this issue can be closed now.

**Comment by copumpkin** [comment]
@sanggusti I don't think this issue can be closed yet, unless I'm misunderstanding. This one is talking about "find me the differences between image X and image Y" or "find common objects between multiple images". The one you link to just runs a batch of images at once, but without the ability to reason about commonalities or differences between them
